src/car_backed_blockstore.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! It can often be time, memory, or disk prohibitive to read large snapshots into a database like [`ParityDb`](crate::db::parity_db::ParityDb).
+//!
+//! This module provides an implementer of [`Blockstore`] that simply wraps a [CAR file](https://ipld.io/specs/transport/car/carv1).
+//! **Note that all operations on this store are blocking**.
+//!
+//! On creation, [`CarBackedBlockstore`] builds an in-memory index of the [`Cid`]s in the file,
+//! and their offsets into that file.
+//!
+//! When a block is requested [`CarBackedBlockstore`] scrolls to that offset, and reads the block, on-demand.
+//!
+//! Writes are currently cached in-memory.
+//!
+//! Performance is pathological with random access.
+//!
+//! # CAR Layout and seeking
+//!
+//! CARs consist of _frames_.
+//! Each frame is a concatenation of the body length as an [`unsigned_varint`], and the _frame body_ itself.
+//! [`unsigned_varint::codec`] can be used to read frames piecewise into memory.
+//!
+//! The first frame's body is a [`CarHeader`] encoded using [`ipld_dagcbor`](serde_ipld_dagcbor).
+//!
+//! Subsequent frame bodies are _blocks_, a concatenation of a [`Cid`] and binary `data`.
+//!
+//! The `offset` in [`BlockLocation`] is the offset of the frame body from the start of the file, illustrated below
+//!
+//! ```text
+//! ┌──────────────┬──────────┐
+//! │  frame       │          │
+//! ├┬────────────┬┤          │
+//! ││ length     ││          │
+//! │┼────────────┼│          │
+//! ││ body       ││          │
+//! │┼┬──────────┬┼┼─┐        │
+//! │││car header│││ │        │
+//! ├┴┴──────────┴┴┤ ▼ length │
+//! │  frame       │          │
+//! ├┬────────────┬┤          │
+//! ││ length     ││          │
+//! │┼────────────┼│          │
+//! ││ body       ││          │
+//! │┼┬──────────┬┼┼─┐        ▼ offset
+//! │││cid       │││ │
+//! ││├──────────┤││ |
+//! │││data      │││ |
+//! ├┴┴──────────┴┴┤ ▼ length
+//! │  frame...    │
+//! ```
+//!
+//! # Future work
+//! - [`fadvise`](https://linux.die.net/man/2/posix_fadvise)-based APIs to pre-fetch parts of the file, to improve random access performance.
+//! - Use an inner [`Blockstore`] for writes.
+//! - Support compressed snapshots.
```

Comment:
If we had a modest frame size for zstd archives - we could have the same mapping to an archive as we have to raw `.car` file, but with an extra computation step + having to decompress a particular frame. Curious what the performance overhead is for an archive that consists of small frames. It seems that the storage overhead is negligible, even if we include a full header with each frame.